core/domain/stats_jobs_one_off.py

```diff
@@ -127,3 +143,2369 @@ def reduce(key, stringified_values):
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [
+            stats_models.StateAnswersModel, stats_models.MigratedAnswerModel]
+
+    @staticmethod
+    def map(item):
+        item.delete()
+        yield (ClearMigratedAnswersJob._DELETE_KEY, 'Deleted answer.')
+
+    @staticmethod
+    def reduce(key, stringified_values):
+        yield 'Deleted %d answer buckets' % len(stringified_values)
+
+
+class PurgeInconsistentAnswersJob(jobs.BaseMapReduceJobManager):
+    """This job deletes all answers stored in
+    stats_models.StateRuleAnswerLogModel which do not or cannot correspond to an
+    existing, accessible exploration.
+    """
+    _AGGREGATION_KEY = 'aggregation_key'
+    _REMOVED_INVALID_HANDLER = 'rem_invalid_handler'
+    _REMOVED_INVALID_FUZZY_RULE = 'rem_invalid_fuzzy_rule'
+    _REMOVED_RULE_SPEC_TOO_LONG = 'rem_rule_spec_too_long'
+    _REMOVED_PERM_DELETED_EXP = 'rem_perm_deleted_exp'
+    _REMOVED_DELETED_EXP = 'rem_deleted_exp'
+    _REMOVED_IMPOSSIBLE_AGE = 'rem_impossible_age'
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [stats_models.StateRuleAnswerLogModel]
+
+    @staticmethod
+    def map(item):
+        item_id = item.id
+
+        if 'submit' not in item_id:
+            yield (PurgeInconsistentAnswersJob._REMOVED_INVALID_HANDLER, {})
+            yield (PurgeInconsistentAnswersJob._AGGREGATION_KEY, {})
+            item.delete()
+            return
+
+        period_idx = item_id.index('.')
+        exp_id = item_id[:period_idx]
+
+        item_id = item_id[period_idx+1:]
+        handler_period_idx = item_id.index('submit') - 1
+
+        item_id = item_id[handler_period_idx+1:]
+        period_idx = item_id.index('.')
+        handler_name = item_id[:period_idx]
+
+        item_id = item_id[period_idx+1:]
+        rule_str = item_id
+
+        if handler_name != 'submit':
+            yield (PurgeInconsistentAnswersJob._REMOVED_INVALID_HANDLER, {})
+            yield (PurgeInconsistentAnswersJob._AGGREGATION_KEY, {})
+            item.delete()
+            return
+
+        if rule_str == 'FuzzyMatches':
+            yield (PurgeInconsistentAnswersJob._REMOVED_INVALID_FUZZY_RULE, {})
+            yield (PurgeInconsistentAnswersJob._AGGREGATION_KEY, {})
+            item.delete()
+            return
+        if rule_str != 'Default' and rule_str[-1] != ')':
+            yield (PurgeInconsistentAnswersJob._REMOVED_RULE_SPEC_TOO_LONG, {})
+            yield (PurgeInconsistentAnswersJob._AGGREGATION_KEY, {})
+            item.delete()
+            return
+
+        exp_model = exp_models.ExplorationModel.get_by_id(exp_id)
+        if not exp_model:
+            yield (PurgeInconsistentAnswersJob._REMOVED_PERM_DELETED_EXP, {})
+            yield (PurgeInconsistentAnswersJob._AGGREGATION_KEY, {})
+            item.delete()
+            return
+        if exp_model.deleted:
+            # TODO(bhenning): Decide if we actually want to omit these answers.
```

Comment:
You actually strongly leaned the opposite direction the last time we talked about this. :) No extra cost assuming all deleted explorations don't have special surprises when performing migrations (many of them probably do, but most of them are probably already handled since we were previously handling these). I actually err on the side of stripping these answers since:

1. We have roughly deterministic migrations of the current set of live explorations
2. Users cannot undelete explorations